### ISSUE TITLE
Support for modules

### ISSUE THREAD

**Issue by aquavitae** [concourse/concourse#595]
We are using concourse to build and deploy about 50 microservices (the number is growing) and have found that we really need a way to compose pipelines so that each project can own it's pipelines, but make use of shared patterns. Rather than implement our own wrapper around fly for this, I want to try to create something that would be useful outside our project and that could be integrated with concourse.  

My hope is that this leads to a discussion that concludes with a practical design that can be implemented.  I have used templating syntax as proposed in #545. See also #417, this is really a concrete proposal for that issue.

So here is my proposal:
1. Modules are partial templates with an additional `vars` block, e.g.:
   
   ``` yaml
   # modules/build.yml
   vars:
     resources:
       source: git-source
       image: built-image
     job:
       name: build
     git_repo: null
     passed: []
   
   resources:
     - name: (( resources.source ))
       type: git
       source:
         repo: (( git.repo ))
   
     - name: (( resources.compiled ))
       type: s3
       ...
   
     - name: (( resources.image ))
       type: docker-image
       ...
   
   jobs:
     - name: (( job.name ))
       plan:
         - get:  git-source
           passed: (( passed ))
   
         - task: build
           file: (( resource.source ))/.ci/build.yml
   
         - put: (( resources.image ))
   ```
2. Pipelines use and configure modules by loading them in a new `modules` block:
   
   ``` yaml
   # pipeline.yml
   modules:
     - name: build
       file: ./modules/build.yml
       vars:
         passed: [ pre-check ]
         git:
           repo: git@github.com:example/example.git
   
   jobs:
     - name: pre-check
       plan:
         - get: (( build.resources.source ))
   
         - task: check
           file: (( build.resources.source ))/.ci/check.yml
   ```

Things to note:
- The `vars` block in a module defines defaults, and can be overridden in the pipeline `modules/vars` block.
- Variables can be used in the module, or in the pipeline namespaced by the name of the module (e.g.  `(( build.resources.source ))`)
- If two modules define identical jobs or resources, only one is created in the pipeline. If modules define different jobs or resources with the same name an error is raised.

As far as implementation is concerned, there are two options.
1.  Modules are an extension of the templating syntax and are resolved by fly before sending the pipeline to the ATC. In this case the syntax for importing a module should probably match the rest of the templating, e.g. something like
   
   ``` yaml
   # pipeline.yml
   (( module build 
       vars:
         ....
   ))
   ```
   
   This is not as pretty as pure yaml but would be fairly simple to implement I think.
2. Modules are a first class citizen and implemented in the ATC along with resources and jobs.  The biggest complexity with this is how the modules would be discovered. Since they sit outside the `jobs` tree and affect the structure of the pipeline, this would suggest that the ATC needs the ability to fetch resources and rebuild pipelines on the fly which doesn't sound like a good idea to me.

**Comment by robdimsdale** [comment]
I would get a lot of value from such a proposal, and I know many others would too.

I believe that in the future the templates will be resolved server-side in the ATC rather than client-side in fly, so we would probably want to find a way to work with that pattern.

That being said, I do wonder how much complexity needs to be handled by concourse, versus how much can be solved by a third-party tool. One could create something that creates pipelines (before you load them into fly) based on what you described without needing to change anything about concourse. That tool could be shared with the community so others would benefit too.

**Comment by aquavitae** [comment]
I realised after I posted, that my comment about templating was a bit misleading - I actually meant to lean on templating syntax, but still resolve client side.  I've given it quite a bit of thought and the approach best I've come up with is for concourse (ATC + fly) to support multiple yaml documents in a single file. Then the modules could be read locally by fly, merged into a single payload and sent to the ATC.  The ATC could then resolve the templating.

**Comment by donaldguy** [comment]
While I certainly wouldn't mind a native solution, I got around this fairly easily with a Makefile to do dependency tracking and a ruby script, to wit,

```ruby
require 'yaml'
require 'tmpdir'

MERGED_PIPELINE = {
  'resource_types' => [],
  'resources' => [],
  'jobs' => [],
  'groups' => {}
}

components = {
   # shared: '.',
   # component_a: 'relative/dir/foo',
   # ... 
}

other_repos = {
  #external_repo: 'ssh://git@github.com:exampleorg/examplerepo.git'
}

# this file is in ci/create_merged_pipeline.yml
repo_root = File.expand_path('../../', __FILE__)

other_repos.each do |name, url|
  dir_name = File.basename(url, '.git')
  potential_repo = File.expand_path("../#{dir_name}", repo_root)
  path = nil

  puts "Including external repo \"#{name}\" - #{url}:"
  if Dir.exists?(potential_repo)
    puts "Using existing checkout at '#{potential_repo}'"
    path = potential_repo
  else
    tmpdir = Dir.mktmpdir(dir_name)
    system("git clone --depth 1 #{url} #{tmpdir}")
    path = tmpdir
  end
  puts

  components[name] = path
end

components.each do |component, component_path|
  component_root = File.expand_path(component_path, repo_root)
  component_pipeline_yaml_path = File.join(component_root, 'ci', 'pipeline.yml')
  next unless File.exists?(component_pipeline_yaml_path)

  component_pipeline_yaml = File.read(component_pipeline_yaml_path)
  #factor out param substitions with something that will read as normal yaml
  component_pipeline_yaml.gsub!('{{','"<concourse_param>')
  component_pipeline_yaml.gsub!('}}','</concourse_param>"')

  begin
    component_pipeline = YAML.load(component_pipeline_yaml)
    MERGED_PIPELINE['resource_types'] += component_pipeline['resource_types'] || []
    MERGED_PIPELINE['resources'] += component_pipeline['resources']
    MERGED_PIPELINE['jobs'] += component_pipeline['jobs'] || []
  rescue Psych::SyntaxError, TypeError => e
    puts "Error: apparent syntax error in #{component_pipeline_yaml_path}"
    p e
    raise e
  end

  unless component_pipeline['jobs'].nil? || component_pipeline['jobs'].empty?
    MERGED_PIPELINE['groups'][component] = component_pipeline['jobs'].map { |i| i['name'] }
  end
end

groups = MERGED_PIPELINE['groups']
# code to merge some groups as
# groups['all'] = MERGED_PIPELINE['jobs'].map { |j| j['name'] }
# groups['combined'] = [ groups.delete(:component_a), groups.delete(:component_a)].flatten

MERGED_PIPELINE['groups'] = groups.each.map do |k,v|
  {
    'name' => k.to_s,
    'jobs' => v
  }
end

MERGED_PIPELINE['groups'].sort! { |a,b| a['name'] <=> b['name'] }

MERGED_PIPELINE_YAML = MERGED_PIPELINE.to_yaml

MERGED_PIPELINE_YAML.gsub!(%r{['"]?<concourse_param>}, '{{')
MERGED_PIPELINE_YAML.gsub!(%r{</concourse_param>['"]?}, '}}')

open(File.join(repo_root, 'ci', '_merged_pipeline.yml'), 'w') do |f|
  f.puts MERGED_PIPELINE_YAML
end
```

**Comment by vito** [comment]
We're moving towards having simple parameterization in the core ( https://github.com/concourse/concourse/issues/545 ) and leaving everything else up to the client. I don't want us to invent another templating language. I'd be interested in seeing patterns and tools explored by third-parties though - ones that cleanly solve actual problems faced day-to-day and may point to improvements we can make in the core.

I'm inclined to close this since we're unlikely to work on it ourselves, but I appreciate the time you took towards making the proposal. I wonder what the next steps should be?

**Comment by aquavitae** [comment]
After trying out a bunch of different approaches for our particular problem, I settled on just being explicit and repeating code.  It's not very DRY, but it is readable and I found most solutions that achieve what I was trying to end up being very complicated to follow and understand.  I think I would rather see more work being done on simplifying or extending the current pipeline syntax to remove some boilerplate. E.g. a way to point to a discoverable set of common resource types, or a way of globally setting some credentials rather than requiring them on each resource.

The one place I am extending the syntax is the use of jinja to make the existing templating more powerful.  The features that I find missing (other than those addressed by #545) are setting default values for variables, and loops.  My approach to building a pipeline is to have a common template used by several projects, and each project provides a yaml file with it's own vars. Some vars are lists (e.g. a list of artifacts which are output), so I use a loop to iterate over those.
